vmhost/contexts/asyncLocal.go

```diff
@@ -326,45 +298,41 @@ func (context *asyncContext) createCallbackInput(
 			OriginalTxHash:       runtime.GetOriginalTxHash(),
 			PrevTxHash:           runtime.GetPrevTxHash(),
 			ReturnCallAfterError: returnWithError,
+			ESDTTransfers:        lastTransfers,
 		},
-		RecipientAddr: context.address,
+		RecipientAddr: caller,
 		Function:      callbackFunction,
 	}
 	context.SetAsyncArgumentsForCallback(contractCallInput, asyncCall, gasAccumulated)
 
-	if isESDTOnCallBack {
-		context.updateContractInputForESDTOnCallback(contractCallInput, esdtFunction, esdtArgs, vmOutput, asyncCall, gasAccumulated)
-	}
-
 	return contractCallInput, nil
 }
 
-func (context *asyncContext) updateContractInputForESDTOnCallback(
-	contractCallInput *vmcommon.ContractCallInput,
-	esdtFunction string,
-	esdtArgs [][]byte,
-	vmOutput *vmcommon.VMOutput,
-	asyncCall *vmhost.AsyncCall,
-	gasAccumulated uint64) {
-
-	oldArgLen := len(contractCallInput.Arguments)
-	oldFunction := contractCallInput.Function
-
-	contractCallInput.Function = esdtFunction
-	contractCallInput.Arguments = make([][]byte, 0, oldArgLen)
-	contractCallInput.Arguments = append(contractCallInput.Arguments, esdtArgs...)
-	contractCallInput.Arguments = append(contractCallInput.Arguments, []byte(oldFunction))
-	contractCallInput.Arguments = append(contractCallInput.Arguments, ReturnCodeToBytes(vmOutput.ReturnCode))
-
-	if len(vmOutput.ReturnData) > 1 {
-		contractCallInput.Arguments = append(contractCallInput.Arguments, vmOutput.ReturnData[1:]...)
+func (context *asyncContext) extractLastTransferToCaller(caller []byte, outputAccounts map[string]*vmcommon.OutputAccount) ([]*vmcommon.ESDTTransfer, []byte) {
+	var lastESDTTransfers *vmcommon.ParsedESDTTransfers
+	var encodedLastTransfers []byte
+	for _, outputAcc := range outputAccounts {
+		if bytes.Equal(outputAcc.Address, caller) {
+			for _, outTransfer := range outputAcc.OutputTransfers {
```

Comment:
you need to check the actual last transfer. so only outputAcc.OutputTransers[len - 1] and if that is an esdt transfer and only if no extra data. and there is no callvalue on VMOutput and no returnData on VMOutput